### ISSUE TITLE
RECURSIVE cte is very slow

### ISSUE THREAD

**Issue by l1t1** [duckdb/duckdb#8700]
### What happens?

a CTE includes 300000 times iteration took 3 minutes. 
postgrsql did the same work in less than 1 second. 

### To Reproduce

```sql
create table cnt as WITH RECURSIVE
   cnt(x) AS (
      SELECT 1
      UNION ALL
      SELECT x+1 FROM cnt
       where x < 300000
) select * from cnt;
```
```text
100% ▕████████████████████████████████████████████████████████████▏
Run Time (s): real 178.950 user 155.891914 sys 91.838243
```

### OS:

aarch64 linux

### DuckDB Version:

0.8.2 dev 3007

### DuckDB Client:

CLI

### Full Name:

lutao

### Affiliation:

home

### Have you tried this on the latest `main` branch?

I have tested with a main build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by Teggy** [comment]
One data point: this query takes about 4 seconds on my machine (Apple M2).  

DuckDB v0.8.2-dev990 188de5715e

**Comment by l1t1** [comment]
could vectorized engine detect the rows in this condition?

**Comment by l1t1** [comment]
an interting finding, `set threads=1` speeds up on windows x64 
```
D WITH RECURSIVE
>    cnt(x) AS (
>       SELECT 1
>         UNION ALL
>       SELECT x+1
>       FROM cnt
>       WHERE x < 300000
> )
> SELECT max(x) FROM cnt;
100% ▕████████████████████████████████████████████████████████████▏
┌────────┐
│ max(x) │
│ int32  │
├────────┤
│ 300000 │
└────────┘
Run Time (s): real 8.173 user 16.224104 sys 1.716011


D WITH RECURSIVE
>    cnt(x) AS (
>       SELECT 1
>       FROM generate_series(1, 1000)    -- place 1000 rows in the working table (formerly: 1 row)
>         UNION ALL
>       SELECT x+1
>       FROM cnt
>       WHERE x < 300000
> )
> SELECT count(x) FROM cnt;
100% ▕████████████████████████████████████████████████████████████▏
┌───────────┐
│ count(x)  │
│   int64   │
├───────────┤
│ 300000000 │
└───────────┘
Run Time (s): real 11.822 user 33.602615 sys 1.528810

D set threads=1;
Run Time (s): real 0.002 user 0.000000 sys 0.000000
D WITH RECURSIVE
>    cnt(x) AS (
>       SELECT 1
>         UNION ALL
>       SELECT x+1
>       FROM cnt
>       WHERE x < 300000
> )
> SELECT max(x) FROM cnt;
100% ▕████████████████████████████████████████████████████████████▏
┌────────┐
│ max(x) │
│ int32  │
├────────┤
│ 300000 │
└────────┘
Run Time (s): real 6.364 user 6.364841 sys 0.000000
D WITH RECURSIVE
>    cnt(x) AS (
>       SELECT 1
>       FROM generate_series(1, 1000)    -- place 1000 rows in the working table (formerly: 1 row)
>         UNION ALL
>       SELECT x+1
>       FROM cnt
>       WHERE x < 300000
> )
> SELECT count(x) FROM cnt;
100% ▕████████████████████████████████████████████████████████████▏
┌───────────┐
│ count(x)  │
│   int64   │
├───────────┤
│ 300000000 │
└───────────┘
Run Time (s): real 8.481 user 8.470854 sys 0.000000
```

**Comment by Teggy** [comment]
> an interting finding, `set threads=1` speeds up on windows x64

I see similar behaviour on mac OS.

**Comment by Reminiscent** [comment]
@Teggy @l1t1 Sorry for the late reply. It takes me some time to go through the push-based execution model code in DuckDB. I just learned all of this knowledge, so feel free to correct me if I'm wrong. 
Thanks for @l1t1 input, I think the main reason why the recursive CTE is very slow is because of the multi-threads. Let me explain why.

I will not explain the push-based execution model here. But if you want to learn more details about the push-based execution model in DuckDB. There are some good materials you can reference from @Mytherin(Thanks to Mark's good sharing!). 
* https://github.com/duckdb/duckdb/issues/1583
* https://github.com/duckdb/duckdb/pull/2393
* [Push-Based Execution in DuckDB by Mark Raasveldt (CWI)](https://www.youtube.com/watch?v=1kDrPgRUuEI)

For the query 
```
WITH RECURSIVE
   cnt(x) AS (
      SELECT 1
      UNION ALL
      SELECT x+1 FROM cnt
       where x < 300000
) select count(*) from cnt;
```
You can treat there are two pipelines here. But the `PhysicalRecursiveCTE` is a little bit special here, because it has `recursive_meta_pipeline` inside:
Pipeline1: `SELECT 1` -> `PhysicalDummyScan(Source) -> PhysicalRecursiveCTE(Sink)`
Pipeline2: `select count(*) from cnt` -> `PhysicalRecursiveCTE(Source) -> PhysicalUngroupedAggregate(Sink)`
`PhysicalRecursiveCTE. recursive_meta_pipeline`: `WITH RECURSIVE cnt(x) AS ( SELECT 1 UNION ALL SELECT x+1 FROM cnt where x < 300000) -> `PhysicalColumnDataScan(Source) -> PhysicalFilter -> PhysicalRecursiveCTE(Sink)`

I would simulate the execution of this query:
1. The `Pipeline1` will execute first. The `PhysicalDummyScan` will generate the value `1`, and then through the `PhysicalRecursiveCTE.Sink`, the value will be pushed to `PhysicalRecursiveCTE.RecursiveCTEState.intermediate_table`. It's no special here.
2. Then, we will handle the `Pipeline2`. At first, it will call the `FetchFromSource -> GetData` to get the data from the `Source` node, here is the `PhysicalRecursiveCTE`. So for the `PhysicalRecursiveCTE::GetData`, it should construct all of the data in the recursive CTE table.
3. As for the process of constructing the recursive CTE table. It's a little bit complex here. So I'll explain it briefly here. Each time, The value(Even if there are several values, like `SELECT 1 FROM generate_series(1, 1000)` is the same) will be put in the `working_table`. And then it will execute the `recursive_meta_pipeline` part to generate the new value which will be put into the `intermediate_table`. After that, the newly generated value in `intermediate_table` will be put into the chunk and passed to the caller function. Then the `intermediate_table` will become the new `working_table` and generate the new value until the recursive part can not meet the requirements(Here is `where x < 300000`).

So I think even if there are several threads, only one thread can handle the `PhysicalRecursiveCTE::GetData`. But there are a lot of time will be wasted because of the multi-threads.

Here are some profilers I want to show to prove my assumption:
If we `set threads = 1`, most of the time will be used to do the true execution.
![image](https://github.com/duckdb/duckdb/assets/41809508/d20dd253-2b08-4ed9-a8c7-fcc385f0f01c)

But if we `set threads = 10`, only a few times of one thread will be used to do the true execution.
![image](https://github.com/duckdb/duckdb/assets/41809508/446f1079-3bb5-491f-81d1-8e5112daae96)

Other threads will spend a lot of time waiting and doing some other small tasks.
![image](https://github.com/duckdb/duckdb/assets/41809508/c2bca42c-d959-4c0f-a081-e44b36ac906f)
 
**TL;DR - If there are pipeline's source operator is  `PhysicalRecursiveCTE`, it can not be run parallel.** 

@Mytherin I am not so sure about my conclusion. Can you help me to take a look? It seems this problem can not be easily solved.

**Comment by Reminiscent** [comment]
@l1t1 Another quick question is how to show the run time for each query. After checking the doc, I only found that the `explain analyze` statement can do that.

**Comment by l1t1** [comment]
`.timer on` at CLI

**Comment by l1t1** [comment]
https://github.com/tableau/hyper-api-samples/issues/101

**Comment by l1t1** [comment]
v0.9.2-dev242 e807b416e8 vs tableauhyperapi         0.0.18161
```
D:\duckdb>duckdb092g
v0.9.2-dev242 e807b416e8
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D .timer on
D create table cnt as WITH RECURSIVE
>    cnt(x) AS (
>       SELECT 1
>       UNION ALL
>       SELECT x+1 FROM cnt
>        where x < 300000
> ) select * from cnt;
Run Time (s): real 9.467 user 20.592132 sys 2.402415
D set threads=1;
Run Time (s): real 0.003 user 0.015600 sys 0.000000
D drop table cnt;
Run Time (s): real 0.001 user 0.000000 sys 0.000000
D create table cnt as WITH RECURSIVE
>    cnt(x) AS (
>       SELECT 1
>       UNION ALL
>       SELECT x+1 FROM cnt
>        where x < 300000
> ) select * from cnt;
Run Time (s): real 6.477 user 6.442841 sys 0.015600

D WITH RECURSIVE
>    cnt(x) AS (
>       SELECT 1
>       FROM generate_series(1, 1000)    -- place 1000 rows in the working table (formerly: 1 row)
>         UNION ALL
>       SELECT x+1
>       FROM cnt
>       WHERE x < 300000
> )
> SELECT * FROM cnt;
│ 300000000 rows │
└────────────────┘
Run Time (s): real 9.256 user 8.533255 sys 0.686404
D set threads=8
D WITH RECURSIVE
>    cnt(x) AS (
>       SELECT 1
>       FROM generate_series(1, 1000)    -- place 1000 rows in the working table (formerly: 1 row)
>         UNION ALL
>       SELECT x+1
>       FROM cnt
>       WHERE x < 300000
> )
> SELECT * FROM cnt;
│ 300000000 rows │
└────────────────┘
Run Time (s): real 13.213 user 26.208168 sys 3.151220
```
hyperapi
```
WITH RECURSIVE
   cnt(x) AS (
      SELECT 1
      UNION ALL
      SELECT x+1 FROM cnt
       where x < 150000
) select count(*) from cnt;

D:\>python temp.py
0.15800905227661133 :  150000

D:\>python temp.py
0.7790446281433105 :  1500000

D:\>python temp.py
6.97939920425415 :  15000000

D:\>pip list |find "hyper"
tableauhyperapi         0.0.18161

WITH RECURSIVE
   cnt(x) AS (
      SELECT 1 from generate_series(1, 1000)
      UNION ALL
      SELECT x+1 FROM cnt
       where x < 150000
) select count(*) from cnt;

D:\>python temp1.py
1.3600778579711914 :  150000000

D:\>python temp1.py
2.2001259326934814 :  300000000
```

**Comment by Vue-data42** [comment]
observe this response



https://stackdev.space/questions/comparison-between-the-performance-of-common-table-expression-and-create-view-statements
